web/src/test/java/org/springframework/security/web/authentication/LoginUrlAuthenticationEntryPointTests.java

```diff
@@ -231,21 +198,58 @@ public void testServerSideRedirectWithForceHttpsRedirectsCurrentRequest()
 		ep.setUseForward(true);
 		ep.setForceHttps(true);
 		ep.afterPropertiesSet();
-		MockHttpServletRequest request = new MockHttpServletRequest();
-		request.setRequestURI("/bigWebApp/some_path");
-		request.setServletPath("/some_path");
-		request.setContextPath("/bigWebApp");
-		request.setScheme("http");
-		request.setServerName("www.example.com");
-		request.setContextPath("/bigWebApp");
-		request.setServerPort(80);
 
+		MockHttpServletRequest request = mockRequest("http", "www.example.com", 80, "/bigWebApp", "/some_path");
 		MockHttpServletResponse response = new MockHttpServletResponse();
 
 		ep.commence(request, response, null);
 		assertThat(response.getRedirectedUrl()).isEqualTo("https://www.example.com/bigWebApp/some_path");
 	}
 
+	@Test
+	public void testRedirectToLoginWithForwardedHeaders() throws Exception {
+		LoginUrlAuthenticationEntryPoint ep = new LoginUrlAuthenticationEntryPoint("/hello");
+		ep.afterPropertiesSet();
+
+		MockHttpServletRequest request = mockRequest("http", "api.example.com", 80, "/app", "/some_path");
+		addForwardedHeaders(request, "https", "www.example.com", 443, "/api");
+
+		MockHttpServletResponse response = new MockHttpServletResponse();
+
+		ep.commence(request, response, null);
+		assertThat(response.getRedirectedUrl()).isEqualTo("https://www.example.com/api/app/hello");
+	}
+
+	@Test
+	public void testRedirectToLoginWithForwardedHeadersNonStandardPort() throws Exception {
+		LoginUrlAuthenticationEntryPoint ep = new LoginUrlAuthenticationEntryPoint("/hello");
+		ep.afterPropertiesSet();
+
+		MockHttpServletRequest request = mockRequest("http", "api.example.com", 80, "/app", "/some_path");
+		addForwardedHeaders(request, "https", "www.example.com", 9443, "/api");
+
+		MockHttpServletResponse response = new MockHttpServletResponse();
+
+		ep.commence(request, response, null);
+		assertThat(response.getRedirectedUrl()).isEqualTo("https://www.example.com:9443/api/app/hello");
+	}
+
+	@Test
+	public void testForwardWithForceSslAndForwardedHeaders() throws Exception {
+		LoginUrlAuthenticationEntryPoint ep = new LoginUrlAuthenticationEntryPoint("/hello");
+		ep.setForceHttps(true);
+		ep.setUseForward(true);
+		ep.afterPropertiesSet();
+
+		MockHttpServletRequest request = mockRequest("http", "api.example.com", 80, "/app", "/some_path");
+		addForwardedHeaders(request, "https", "www.example.com", 9443, "/api");
+
+		MockHttpServletResponse response = new MockHttpServletResponse();
+
+		ep.commence(request, response, null);
+		assertThat(response.getRedirectedUrl()).isEqualTo("https://www.example.com:9443/api/app/hello");
```

Comment:
This fails currently.  The actual result will be "https://www.example.com/api/app/some_path".  This is related to my second bullet point in the pull request description.  Not sure what the correct behavior would be here.